File: examples/injectable-twitter-server/java/src/main/java/com/twitter/finatra/example/Subscriber.java
package com.twitter.finatra.example;

import java.util.Optional;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import javax.inject.Inject;
import javax.inject.Singleton;

import scala.runtime.BoxedUnit;

import com.twitter.inject.annotations.Flag;
import com.twitter.util.AbstractCloseAwaitably;
import com.twitter.util.Duration;
import com.twitter.util.Function0;
import com.twitter.util.Future;
import com.twitter.util.NullTimerTask$;
import com.twitter.util.ScheduledThreadPoolTimer;
import com.twitter.util.Time;
import com.twitter.util.TimerTask;
import com.twitter.util.logging.Logger;

@Singleton
public class Subscriber extends AbstractCloseAwaitably {
  private final Logger logger = Logger.apply(Subscriber.class);

  private final Queue queue;
  private final int maxRead;
  private final AtomicInteger numRead;
  private final ScheduledThreadPoolTimer timer;
  private final AtomicReference<TimerTask> task;

  @Inject
  public Subscriber(
      Queue queue,
      @Flag("subscriber.max.read") int maxRead) {
    this.queue = queue;
    this.maxRead = maxRead;
    this.numRead = new AtomicInteger();
    this.timer = new ScheduledThreadPoolTimer(1, "SubscriberPool", false);
    this.task = new AtomicReference<>(NullTimerTask$.MODULE$);
  }

  @Override
  public Future<BoxedUnit> onClose(Time deadline) {
    TimerTask currentTask = task.get();
    if (!NullTimerTask$.MODULE$.equals(currentTask)) {
      logger.info("Subscriber has stopped reading from queue.");
      return currentTask.close(deadline);
    } else {
      logger.info("Subscriber closed.");
      return Future.Done();
    }
  }

  public int readCount() {
    return numRead.get();
  }

  /**
   * Begin the scheduled task of reading items from the queue.
   */
  public void start() {
    logger.info("Starting Subscriber");
    task.compareAndSet(
        NullTimerTask$.MODULE$,
        timer.schedule(
          Time.now().plus(Duration.fromSeconds(2)),
          Duration.fromSeconds(1),
          new Function0<BoxedUnit>() {
          @Override
          public BoxedUnit apply() {
            return read();
          }
        })
    );
  }

  private BoxedUnit read() {
    logger.info("Subscriber#read()");
    Optional<String> str = queue.poll();
    str.ifPresent(s -> {
      numRead.incrementAndGet();
      logger.info("Reading " + s + " value from the queue. Read: " + numRead.get() + ".");
    });
    return BoxedUnit.UNIT;
  }

  @Override
  public boolean isReady(CanAwait permit) {
    return numRead.get() >= this.maxRead;
  }
}

File: examples/injectable-twitter-server/scala/src/main/scala/com/twitter/finatra/example/Subscriber.scala
package com.twitter.finatra.example

import com.twitter.conversions.DurationOps._
import com.twitter.inject.Logging
import com.twitter.inject.annotations.Flag
import com.twitter.util._
import java.util.concurrent.atomic.{AtomicInteger, AtomicReference}
import javax.inject.{Inject, Singleton}

@Singleton
class Subscriber @Inject() (
  queue: Queue,
  @Flag("subscriber.max.read") maxRead: Int)
    extends ClosableOnce
    with CloseOnceAwaitably
    with Logging {

  private[this] val numRead: AtomicInteger = new AtomicInteger(0)
  private[this] val PoolTimer: ScheduledThreadPoolTimer =
    new ScheduledThreadPoolTimer(1, "SubscriberPool", false)
  private[this] val task: AtomicReference[TimerTask] = new AtomicReference[TimerTask](NullTimerTask)

  def closeOnce(deadline: Time): Future[Unit] = {
    task.get() match {
      case NullTimerTask =>
        info("Subscriber closed.")
        Future.Done
      case task: TimerTask =>
        info("Subscriber has stopped reading from queue.")
        task.close(deadline)
    }
  }

  def readCount: Int = numRead.get()

  def start(): Unit = {
    info("Starting Subscriber.")
    task.compareAndSet(
      NullTimerTask, // should be null in order to start
      PoolTimer.schedule(Time.now.plus(2.seconds), 1.seconds)(read())
    )
  }

  private[this] def read(): Unit = {
    info("Subscriber#read()")
    queue.poll match {
      case Some(value) =>
        numRead.incrementAndGet()
        info("Reading " + value + " value from the queue. Read: " + numRead.get + ".")
      case _ =>
      // do nothing
    }
  }

  override def isReady(implicit permit: Awaitable.CanAwait): Boolean = numRead.get >= maxRead
}
